### ISSUE TITLE
`azurerm_role_assignment` forces replace after import

### ISSUE THREAD

**Issue by stephenhogg-des** [hashicorp/terraform-provider-azurerm#23487]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.4.6

### AzureRM Provider Version

3.55.0

### Affected Resource(s)/Data Source(s)

azurerm_role_assignment

### Terraform Configuration Files

```hcl
resource "azurerm_role_assignment" "test" {
  scope                = data.azurerm_resource_group.rg-fs.id
  role_definition_name = "User Access Administrator"
  principal_id         = data.azuread_group.common-resourcecontributor.object_id
}

data "azuread_group" "common-resourcecontributor" { display_name = "group_name" }

data "azurerm_resource_group" "rg-fs" { name = "rg_name" }
```

### Debug Output/Panic Output

```shell
Terraform will perform the following actions:

  # azurerm_role_assignment.test must be replaced
-/+ resource "azurerm_role_assignment" "test" {
      ~ id                               = "/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED" -> (known after apply)
      ~ name                             = "REDACTED" -> (known after apply)
      ~ principal_type                   = "Group" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/REDACTED/providers/Microsoft.Authorization/roleDefinitions/REDACTED" -> (known after apply)
      ~ scope                            = "/subscriptions/REDACTED/resourcegroups/REDACTED" -> "/subscriptions/REDACTED/resourceGroups/REDACTED" # forces replacement
      + skip_service_principal_aad_check = (known after apply)
        # (2 unchanged attributes hidden)

      - timeouts {}
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```

### Expected Behaviour

`azurerm_role_assignment` should not be replaced due to forced lowercase of scope. This line:

```      
~ scope                            = "/subscriptions/REDACTED/resourcegroups/REDACTED" -> "/subscriptions/REDACTED/resourceGroups/REDACTED" # forces replacement
```

shouldn't be happening because the string imported contained the substring `resourceGroups` but it got lowercased by the provider during import

### Actual Behaviour

The provider detects there is a difference in casing between the imported scope and the reported one and attempts to do a replacement, because the scope got lowercased during import.

### Steps to Reproduce

1. `terraform import azurerm_role_assignment.test INSERT_CORRECTLY_CASED_ID_HERE`
2. `terraform apply`

**Comment by wuxu92** [comment]
Hi @stephenhogg-des Thanks for bringing up this issue. Although I am unable to replicate the problem with the given configuration, could you please obtain the DEBUG log by running `TF_LOG=debug tf import azurerm_role_assignment.test <INSERT_CORRECTLY_CASED_ID_HERE>`? This will help us determine if it is caused by an API issue.

**Comment by stephenhogg-des** [comment]
Hi @wuxu92 sure thing - see below. You'll see on the first line that the CLI args include `resourceGroups` in the ID that gets imported but when the object is created at the end of the log the `.scope` field shows as `resourcegroups`. This is what's causing the problem. Please do let me know if anything else is needed.

```
2023-10-11T07:58:17.991+1000 [INFO]  CLI args: []string{"terraform", "import", "module.REDACTED.azurerm_role_assignment.common-resource-contributor-uaa-for-rg", 
"/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED"}
2023-10-11T07:58:21.444+1000 [DEBUG] ProviderTransformer: "module.REDACTED.azurerm_role_assignment.common-resource-contributor-uaa-for-rg (expand)" (*terraform.nodeExpandPlannableResource) needs provider["registry.terraform.io/hashicorp/azurerm"]
2023-10-11T07:58:35.789+1000 [DEBUG] ReferenceTransformer: "module.REDACTED.azurerm_role_assignment.common-resource-contributor-uaa-for-rg (import id 
\"/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED\")" references: []
[0m[1mmodule.REDACTED.azurerm_role_assignment.common-resource-contributor-uaa-for-rg: Importing from ID "/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED"...[0m
[0m[1m[32mmodule.REDACTED.azurerm_role_assignment.common-resource-contributor-uaa-for-rg: Import prepared![0m
[0m[1mmodule.REDACTED.azurerm_role_assignment.common-resource-contributor-uaa-for-rg: Refreshing state... [id=/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED][0m
2023-10-11T07:58:36.827+1000 [DEBUG] provider.terraform-provider-azurerm_v3.55.0_x5.exe: AzureRM Request: 
GET //subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED?api-version=2020-04-01-preview HTTP/1.1
Host: management.azure.com
User-Agent: Go/go1.19.3 (amd64-windows) go-autorest/v14.2.1 Azure-SDK-For-Go/v66.0.0 authorization/2020-04-01-preview HashiCorp Terraform/1.4.6 (+https://www.terraform.io) Terraform Plugin SDK/2.10.1 terraform-provider-azurerm/dev pid-222c6c49-1b0a-5959-a213-6608f9eb8820
X-Ms-Correlation-Request-Id: REDACTED
Accept-Encoding: gzip: timestamp=2023-10-11T07:58:36.826+1000

2023-10-11T07:58:36.830+1000 [INFO]  provider.terraform-provider-azuread_v2.36.0_x5.exe: 2023/10/11 07:58:36 [DEBUG] ============================ Begin AzureAD Response ===========================
GET https://graph.microsoft.com/beta/groups/07c1b7d7-0481-4e1d-a529-b5590d494999/owners?%24select=id
Request ID: REDACTED

HTTP/1.1 200 OK
Transfer-Encoding: chunked
Cache-Control: no-cache
Client-Request-Id: REDACTED
Content-Type: application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8
Date: Tue, 10 Oct 2023 21:58:36 GMT
Odata-Version: 4.0
Request-Id: ac789e59-34a2-4366-9f23-678512e13c73
Strict-Transport-Security: max-age=31536000
Vary: Accept-Encoding
X-Ms-Ags-Diagnostic: {"ServerInfo":{"DataCenter":"Australia East","Slice":"E","Ring":"5","ScaleUnit":"000","RoleInstance":"SY3PEPF000043FC"}}
X-Ms-Resource-Unit: 1

5f
{"@odata.context":"https://graph.microsoft.com/beta/$metadata#directoryObjects(id)","value":[]}
0


============================= End AzureAD Response ============================: timestamp=2023-10-11T07:58:36.829+1000

2023-10-11T07:58:38.522+1000 [WARN]  Provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new value for module.REDACTED.azurerm_role_assignment.common-resource-contributor-uaa-for-rg during refresh.
      - .principal_type: was null, but now cty.StringVal("Group")
      - .role_definition_id: was null, but now cty.StringVal("/subscriptions/REDACTED/providers/Microsoft.Authorization/roleDefinitions/18d7d88d-d35e-4fb5-a5c3-7773c20a72d9")
      - .role_definition_name: was null, but now cty.StringVal("User Access Administrator")
      - .condition: was null, but now cty.StringVal("")
      - .delegated_managed_identity_resource_id: was null, but now cty.StringVal("")
      - .name: was null, but now cty.StringVal("REDACTED")
      - .principal_id: was null, but now cty.StringVal("REDACTED")
      - .scope: was null, but now cty.StringVal("/subscriptions/REDACTED/resourcegroups/REDACTED")
      - .condition_version: was null, but now cty.StringVal("")
      - .description: was null, but now cty.StringVal("")
```

**Comment by wuxu92** [comment]
@stephenhogg-des It looks like the DEBUG LOG redacted the GET Request of the role assignment resource. Could you please find out the log like `2023-10-11T10:16:59.092+0800 [DEBUG] provider.terraform-provider-azurerm_v3.71.0_x5: AzureRM Response for h[64/1858]nagement.azure.com//subscriptions/xxx/resourceGroups/test-rg/providers/Microsoft.Authorization/roleAssignments/xxx?api-version=2020-04-01-preview:` and share the masked response of it?

**Comment by stephenhogg-des** [comment]
Apologies @wuxu92 , here it is:

```
2023-10-11T14:34:43.956+1000 [DEBUG] provider.terraform-provider-azurerm_v3.55.0_x5.exe: AzureRM Response for 
https://management.azure.com//subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED?api-version=2020-04-01-preview: 
HTTP/2.0 200 OK
Content-Length: 912
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
Date: Wed, 11 Oct 2023 04:34:43 GMT
Expires: -1
Pragma: no-cache
Set-Cookie: x-ms-gateway-slice=Production; path=/; secure; samesite=none; httponly
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Cache: CONFIG_NOCACHE
X-Content-Type-Options: nosniff
X-Ms-Correlation-Request-Id: REDACTED
X-Ms-Ratelimit-Remaining-Subscription-Reads: 11996
X-Ms-Request-Id: REDACTED
X-Ms-Routing-Request-Id: AUSTRALIAEAST:20231011T043443Z:a01dfd96-369d-48e3-9b21-3d2989762931
X-Msedge-Ref: Ref A: DD35D729FE554970B69A93C29F20C56B Ref B: MAA201060515021 Ref C: 2023-10-11T04:34:43Z

{"properties":{"roleDefinitionId":"/subscriptions/REDACTED/providers/Microsoft.Authorization/roleDefinitions/18d7d88d-d35e-4fb5-a5c3-7773c20a72d9","principalId":"REDACTED","principalType":"Group","scope":"/subscriptions/REDACTED/resou
rcegroups/REDACTED","condition":null,"conditionVersion":null,"createdOn":"2023-08-23T02:01:36.9180599Z","updatedOn":"2023-10-05T01:14:28.6866981Z","createdBy":"REDACTED","updatedBy":"REDACTED","delegatedManagedIdentityResourceId":null,"description":null
},"id":"/subscriptions/REDACTED/resourcegroups/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED","type":"Microsoft.Authorization/roleAssignments","name":"REDACTED"}: timestamp=2023-10-11T14:34:43.955+1000
```

I went and ran `az graph` as well to check, and it looks like it's just the Azure API returning `resourcegroups`, which the provider then wants to replace with `resourceGroups`. Hope this helps.

**Comment by wuxu92** [comment]
Thanks for your information. Then this looks like an API issue. but is it possible that the `scope` field was set to a lower-cased `resourcegroup` when creating the roleAssignment? It is an expected behavior if so.

**Comment by stephenhogg-des** [comment]
The problem is regardless of which way I import it, with either `resourceGroups` or `resourcegroups` the state file winds up with a scope that is `resourcegroups`, which it then tries to overwrite on the next `apply`.

**Comment by wuxu92** [comment]
The `scope` field of the state file was written with the value extracted from the API response, say this part of the response payload  `"principalType":"Group", "scope":"/subscriptions/REDACTED/resourcegroups/REDACTED"`. It does not relate to the way you import it but to the API response payload.

**Comment by stephenhogg-des** [comment]
There's another problem where an import will fail entirely if the role assignment ID has `RoleAssignments` instead of `roleAssignments` in it that has the same root cause - see below:

```
[31mÔöé[0m [0m[1m[31mError: [0m[0m[1mparsing Resource ID "/subscriptions/REDACTED/resourcegroups/REDACTED/providers/Microsoft.Authorization/RoleAssignments/REDACTED": ID was missing the `roleAssignments` element[0m
```

I know that ID is completely valid (the Azure API provided it to me), but the provider says it isn't. Both of these problems come down to casing restrictions that exist in the provider but shouldn't because they don't exist in the API. So given all that, how do we fix the provider? I'm happy to contribute a fix with a bit of guidance as to what needs to change.

**Comment by wuxu92** [comment]
@stephenhogg-des AFAIK. This behavior is by design. The ID in Terraform is case-sensitive, which differs from the Azure API. As you say `Azure API provided it to me` Could you please share which API returns this value? We need to address the root cause of the issue.

**Comment by stephenhogg-des** [comment]
Sure - I got it from a call to `az graph`. The specific command was:

```
az graph query -q "authorizationresources | where type =~ 'microsoft.authorization/roleassignments' | where properties.scope endswith 'REDACTED' | where properties.principalId == 'REDACTED' and properties.roleDefinitionId == '/providers/Microsoft.Authorization/RoleDefinitions/18d7d88d-d35e-4fb5-a5c3-7773c20a72d9' | project id" --first 1000 --query data[0].id
```

**Comment by stephenhogg-des** [comment]
This issue appears to be saying the same thing - https://github.com/hashicorp/terraform-provider-azurerm/issues/23519

**Comment by stephenhogg-des** [comment]
And another one - https://github.com/hashicorp/terraform-provider-azurerm/issues/14625 

Sorry for comment spam, was just looking through issues and spotted a few

**Comment by wuxu92** [comment]
The API/`az` could return a case-insensitive result which is not accepted by `terraform import`. You need to change it to the right case.

**Comment by stephenhogg-des** [comment]
Ok - a few questions, if it's ok?

- There is currently case insensitivity for the `role_definition_id` and `role_definition_name` fields, is there any particular reason that's ok and case insensitivity for the `scope` field is not? I notice that both the `role_definition_id` and `scope` can be acquired from a case insensitive API
- If I were to change it to the right case as you say so that in practice there is case insensitivity, how would I go about it?

**Comment by wuxu92** [comment]
@stephenhogg-des 

1. `role_definision_id` and `role_definision_name` are defined with `DiffSuppressFunc: suppress.CaseDifference`, So they are case insensitive. But the `scope` field is not defined with suppress. And by the latest provider guideline, we don't add `CaseDifference` suppress to fields anymore.

https://github.com/hashicorp/terraform-provider-azurerm/blob/bc4f55fc37deca181ec567278ca9c31a5e95d9da/internal/services/authorization/role_assignment_resource.go#L77-L94

2. Do you mean how to update the `func DestructRoleAssignmentId` logic to make its case insensitive? I do not think it's a good way to achieve. If the `scope` field caused the issue, then we have to provide the value (with the right case) of the `scope`.

**Comment by stephenhogg-des** [comment]
Hi again @wuxu92 - thanks for your answers. As regards the second question, I just am asking how to create a PR to fix this issue. Any advice would be greatly appreciated, then I will go and do it :)

**Comment by wuxu92** [comment]
@stephenhogg-des Thanks! There is a way to implement the PR: adding a `DiffSuppressFunc` just like the `role_definition_name` field. This is not recommended anymore, But you can have a try with a solid reason in your PR description.

**Comment by tombuildsstuff** [comment]
@stephenhogg-des @wuxu92 unfortunately this isn't a case where a DiffSuppressFunc should be added, so we wouldn't accept a PR for that at this time.

Ultimately this limitation is coming from inconsistencies within the Azure API - the ARM Team have stated that the Azure API should accept items _case insensitively_ and return these _case sensitively_ notably including Resource ID segments which should also be camelCase'd (e.g. `subscriptions`, `resourceGroups`, `webApps`) - however many Azure API's do not conform with this behaviour. Whilst adding a `DiffSuppressFunc` seems like a quick solution to this issue, it actually exacerbates the issue in the long run, by enabling users to unintentionally import Resources (and configure fields) which don't match the casing we're expecting, which whilst being (typically) a silent warning today - are surfaced as errors in the next version of the Terraform protocol (e.g. `Diff didn't match the apply`).

As mentioned above in some cases the casing of the Key component of the Resource ID (e.g. `resourceGroups` in the key value pair of `/resourceGroups/name`) matters - we've seen instances of this where Resources are provisioned in the Azure API but aren't actually usable, which also applies to fields within the API Payload itself.

Ultimately this is why Terraform validates the casing matches what we're expecting here - and is why adding a `DiffSuppressFunc` will actually just exacerbate this issue down the line - and is why (unfortunately) we won't accept a PR adding a `DiffSuppressFunc` to this field.

---

Whilst I appreciate that's not going to help your scenario immediately - ultimately this is something we're working to resolve and is a large part of what moving from `Azure/azure-sdk-for-go` to `hashicorp/go-azure-sdk` enables - however unfortunately we need to be fully migrated across to be able to proceed with some of the mitigation here (else fixing this in one SDK used by one resource but not in another will cause knock-on issues).

In the interim however it should be possible to fix this by conditionally recasing the value returned from the API Response as we do in certain other types of resources - @wuxu92 would you mind taking a look into that one?

Thanks!

**Comment by Taha-cmd** [comment]
Hi,
this was supposed to be fixed in v3.81.0, which is now released. The release notes didn't mention this fix. Also, I just tried v3.81.0 locally and the problem is still present. Will this be released in the next version? Or is the fix not working as intended? Just asking to make sure that this doesn't get forgotten, because the PR is already merged and the issue is closed.

**Comment by Taha-cmd** [comment]
@wuxu92 sorry for the late answer.
```
Terraform v1.6.0
on windows_amd64
+ provider registry.terraform.io/hashicorp/azuread v2.41.0
+ provider registry.terraform.io/hashicorp/azurerm v3.81.0
```

```

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create
  ~ update in-place
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # azuread_application.all["key1"] will be updated in-place
  # (imported from "<REDACTED>")
  ~ resource "azuread_application" "all" {
      <REDACTED>
    }

  # azuread_application.all["key2"] will be updated in-place
  # (imported from "<REDACTED>")
  ~ resource "azuread_application" "all" {
      <REDACTED>
    }

  # azuread_service_principal.all["key1"] will be updated in-place
  # (imported from "<REDACTED>")
  ~ resource "azuread_service_principal" "all" {
      <REDACTED>
    }

  # azuread_service_principal.all["key2"] will be updated in-place
  # (imported from "<REDACTED>")
  ~ resource "azuread_service_principal" "all" {
      <REDACTED>
    }

  # azurerm_container_registry.this will be updated in-place
  # (imported from "<REDACTED>")
  ~ resource "azurerm_container_registry" "this" {
        <REDACTED>
    }

  # azurerm_role_assignment.acr_pullers["KEY"] must be replaced
  # (imported from "/subscriptions/<REDACTED>/resourceGroups/<REDACTED>/providers/Microsoft.ContainerRegistry/registries/<REDACTED>/providers/Microsoft.Authorization/roleAssignments/<REDACTED>")
  # Warning: this will destroy the imported resource
-/+ resource "azurerm_role_assignment" "acr_pullers" {
      ~ id                               = "/subscriptions/<REDACTED>/resourceGroups/<REDACTED>/providers/Microsoft.ContainerRegistry/registries/<REDACTED>/providers/Microsoft.Authorization/roleAssignments/<REDACTED>" -> (known after apply)
      ~ name                             = "<REDACTED>" -> (known after apply)
      ~ principal_id                     = "<REDACTED>" -> (known after apply)
      ~ principal_type                   = "ServicePrincipal" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/<REDACTED>/providers/Microsoft.Authorization/roleDefinitions/<REDACTED>" -> (known after apply)
        role_definition_name             = "AcrPull"
      ~ scope                            = "/subscriptions/<REDACTED>/resourcegroups/<REDACTED>/providers/Microsoft.ContainerRegistry/registries/<REDACTED>" -> "/subscriptions/<REDACTED>/resourceGroups/<REDACTED>/providers/Microsoft.ContainerRegistry/registries/<REDACTED>" # forces replacement
      + skip_service_principal_aad_check = true
    }

Plan: 6 to import, 3 to add, 5 to change, 1 to destroy.

```
